pslab/bus/spi.py

```diff
@@ -0,0 +1,505 @@
+"""Control the PSLab's SPI bus and devices connected on the bus.
+
+Examples
+--------
+Set SPI bus speed to 200 kbit/s:
+
+>>> from pslab.bus.spi import SPIMaster, SPISlave
+>>> bus = SPIMaster()
+>>> bus.set_parameters(primary_prescaler=0, secondary_prescaler=3) # 64e6/(64*5)
+
+Set SPI bus to mode 3 (1,1):
+
+>>> bus.set_parameters(0, 3, CKE=0, CKP=1)
+
+Transfer a random byte over SPI:
+
+>>> slave = SPISlave()
+>>> slave.transfer8(0x55)
+0
+"""
+
+from typing import List
+
+import pslab.protocol as CP
+from pslab.serial_handler import SerialHandler
+
+__all__ = (
+    "SPIMaster",
+    "SPISlave",
+)
+# Default values, refer pslab-firmware.
+PPRE = 0
+SPRE = 2
+# SPI mode 0 (0,0)
+CKE = 1  # Clock Phase 0 | Clock Edge 1
+CKP = 0  # Clock Polarity 0
+SMP = 1
```

Comment:
Perhaps these could be _underscored? Or are they intended to be modified by the user?

Reply:
I thought, it is not needed to prefix with underscore since we have `__all__`.

It's all constants, since all of these in capital letters.

Reply:
I think `__all__` only affects the behavior of `from pslab.bus.spi import *`. If someone does `from pslab.bus import spi` they will be visible.

Reply:
Yeah, I will change it :+1:

Reply:
Could I change SPIPrimitive too, since it is not intended for user?

Reply:
But we have to do something for `spi.List` and `spi.SerialHandler` too.

Reply:
Some namespace pollution is all but inevitable, let's just do what we can to minimize it. No need to do anything about imported names.

Reply:
So the changes are:
```
i2c.I2CPrimitive -> i2c._I2CPrimitive
spi.SPIPrimitive -> spi._SPIPrimitive
spi.PPRE -> spi._PPRE
spi.SPRE -> spi._SPRE
spi.CKE -> spi._CKE
spi.CKP -> spi._CKP
spi.SMP -> spi._SMP
```